### ISSUE TITLE
design `upload discovery` tus extension

### ISSUE THREAD

**Issue by butonic** [owncloud/ocis#1298]
To save clients from saving the upload id they would like to be able to discover existing upload ids using the path and a checksum.

This could be designed as a tus extension.

The client will send a post with the metadata and the server should be able to return either a 100-continue or a redirect the existing upload location.

cc @PVince81 @TheOneRing

**Comment by micbar** [comment]
@TheOneRing @butonic which end user value does that bring?

**Comment by TheOneRing** [comment]
By default tus doesn't not use chunks, so if you upload a 100GiB, with `creation-with-upload`, and the upload fails, we never received an upload location to continue and have to start all over again.

We currently work around this issue by using ca 100MiB chunks, but ideally we could roll an id for `creation-with-upload`.

**Comment by michaelstingl** [comment]
> We currently work around this issue by using ca 100MiB chunks

Backend now announces 10 MiB:

https://github.com/owncloud/ocis/blob/b892a9ab286f85a8db7b61164ff234131443ac50/docs/apis/http/tus_upload.md?plain=1#L20-L44

**Comment by michaelstingl** [comment]
> but ideally we could roll an id for `creation-with-upload`.

This is the _Upload Tag extension_ in `tus:1.1+`
- https://github.com/tus/tus-resumable-upload-protocol/pull/163

For this approach, clients roll an id (Upload Tag), but need to remember it to query information about the incomplete upload later.

This issue is about an approach, where clients wouldn't need to persist the id (Upload Tag), but instead use the path and a checksum.

> To save clients from saving the upload id they would like to be able to discover existing upload ids using the path and a checksum.

**Comment by micbar** [comment]
@TheOneRing @michaelstingl 
Let me summarize:

- The clients want to be able to resume uploads and therefore without "remembering" the upload ID
- The clients want to be able to recover from a failed `creation-with-upload` request.

Like @michaelstingl wrote, we have the default of 10MB for the chunk size these days. I am a bit hesitant to add "too much extra stuff" in the light of TUS becoming a standard protocol for vendors like apple.

https://tus.io/blog/2023/08/09/resumable-uploads-ietf

## Parts of the IETF proposal

![1718088693313](https://github.com/owncloud/ocis/assets/4378513/516ed52d-3b8f-411d-b3c2-c5df6dc17e46)

`create` and `creation-with-upload` will be part of that. Bottom line from my POV is, that we should not go beyond that.

**Comment by TheOneRing** [comment]
> @TheOneRing @michaelstingl Let me summarize:
> 
>     * The clients want to be able to resume uploads without "remembering" the upload ID

The client wants to be able to suggest the upload ID, if we only get the upload ID as a response to the first request, we can't resume anything if that request failed.

But I have no strong opinion on this, I just didn't want to see it closed because "Old"

> 
>     * The clients want to be able to recover from a failed `creation-with-upload` request.
> 
> 
> Like @michaelstingl wrote, we have the default of 10MB for the chunk size these days. I am a bit hesitant to add "too much extra stuff" in the light of TUS becoming a standard protocol for vendors like apple.
> 
> https://tus.io/blog/2023/08/09/resumable-uploads-ietf
> ## Parts of the IETF proposal
> 
> ![1718088693313](https://private-user-images.githubusercontent.com/4378513/338469518-516ed52d-3b8f-411d-b3c2-c5df6dc17e46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgwOTQxMjksIm5iZiI6MTcxODA5MzgyOSwicGF0aCI6Ii80Mzc4NTEzLzMzODQ2OTUxOC01MTZlZDUyZC0zYjhmLTQxMWQtYjNjMi1jNWRmNmRjMTdlNDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTFUMDgxNzA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDQ3YTcwNmZjMmZjZjcwNDIwNWZiZjlkNWM0MDdkYWI1NmI4ZThmODNlMGNmOWJkMzgxMGYzMWEzOWFkZmU4YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.V1kNUYtKX0ExvPEWfwcE9FCAq4gRuaZ0Jt9c-jIPyw4)
> 
> `create` and `creation-with-upload` will be part of that. Bottom line from my POV is, that we should not go beyond that.

**Comment by felix-schwarz** [comment]
@micbar @TheOneRing IIRC this issue originated from an early discussion - which eventually culminated in me proposing `Upload-Token` to the tus team and working with them to add it to the spec.

While - technically - `Upload-Token` is a different thing I'm not sure an issue of its own exists for it here.

With regards to the iOS client, supporting `Upload-Token` can
- help to _greatly_ reduce the complexity of the tus implementation in the iOS app, including error handling
- reduce the number of requests sent by the app to the server (unless an error occurs: from one per upload rather than one per 10 MB)
- thereby help avoid a pattern for background transfers that iOS `NSURLSession` can penalize with extreme delays sending background HTTP requests (even if the app is in the foreground) (delays by iOS in sending of scheduled requests by several minutes is not unheard of there)

In short: it would elevate the possible app experience.

That vendors like Apple have started adding basic tus support to their OS is a good point, but it will likely take a couple more releases (years) until we get full-fledged support - and then a few more years until that OS version is old enough to become the minimum required version for running the app. So there's still a long time to go until the iOS app can _depend_ on OS support for tus.
